### PR TITLE
refactor subnet ID

### PR DESCRIPTION
**Pull request by adlrocha**
Propagates the refactor done in Rust here: https://github.com/consensus-shipyard/ipc-actors/pull/109/files

### CODE REVIEW

File: sdk/subnetid.go [hunk 1]
```diff
@@ -1,165 +1,211 @@
...
+	Children []address.Address
 }
 
-var id0, _ = address.NewIDAddress(0)
-
 const (
-	RootStr          = "/root"
+	RootPrefix       = "r"
 	SubnetSeparator  = "/"
 	UndefStr         = ""
```
Comment (dnkolegov) [inline sdk/subnetid.go:23]: This const is not used.

File: sdk/subnetid.go [hunk 2]
```diff
@@ -1,165 +1,211 @@
...
+	return len(id.Children) == 0
+}
+
+func (id SubnetID) ChainID() uint64 {
+	if id.IsRoot() {
+		return id.Root
+	}
+
+	h := fnv.New64a()
+	h.Write([]byte(id.String()))
```
Comment (dnkolegov) [inline sdk/subnetid.go:69]: Unhandled error. 
Can we just ignore it by adding `// nolint ` ?

File: sdk/subnetid.go [hunk 3]
```diff
@@ -2,63 +2,117 @@ package sdk
...
 	return id.String()
 }
 
-func NewSubnetIDFromString(addr string) (SubnetID, error) {
-	var out SubnetID
-	if addr == RootSubnet.String() {
-		out = RootSubnet
-		return out, nil
+// Equal checks if two subnet IDs are equal.
+func (id SubnetID) Equal(other SubnetID) bool {
```
[the remark this answers is not in the document]
Comment (adlrocha) [reply sdk/subnetid.go:43]: @ranchalp, any strong reason behind this? If not we should change it. I inherited this from Alex's initial implementation.

File: sdk/subnetid.go (hunk 1 above)
Comment (adlrocha) [reply sdk/subnetid.go:23]: Yes it is, as part of the `String()` implementation

File: sdk/subnetid.go (hunk 2 above)
Comment (adlrocha) [reply sdk/subnetid.go:69]: Good catch, updating it.

File: sdk/subnetid.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ranchalp) [reply sdk/subnetid.go:43]: No strong reason AFAIR, feel free to change it.

**Pull request merged**